stan/math/opencl/kernels/tridiagonalization.hpp

```diff
@@ -0,0 +1,326 @@
+             step /= REDUCTION_STEP_SIZE) {
+          if (lid < step) {
+            for (int i = 1; i < REDUCTION_STEP_SIZE; i++) {
+              res_loc[lid] += res_loc[lid + step * i];
+            }
+          }
+          barrier(CLK_LOCAL_MEM_FENCE);
+        }
+        if (lid == 0) {
+          V[V_rows * (j + 1) + i + j] = res_loc[lid];
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+      }
+    }
+    // \cond
+);
+// \endcond
+
+// \cond
+const char* tridiagonalization_v3_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Third part in constructing vector v: v-=0.5*(v^T*u)*u, where u is the
+     * householder vector.
+     * @param[in,out] P packed matrix being constructed
+     * @param[in,out] V matrix V
+     * @param[out] q q
+     * @param P_rows Number of rows of the packed matrix
+     * @param V_rows Number of rows of the matrix V
+     * @param k Index of the householder vector in the block to create
+     * @param j Start column of the block to work on
+     */
+    __kernel void tridiagonalization_v3(
+        __global double* P, __global double* V, __global double* q,
+        const int P_rows, const int V_rows, const int k, const int j) {
+      const int lid = get_local_id(0);
+      const int gid = get_global_id(0);
+      const int gsize = get_global_size(0);
+      const int lsize = get_local_size(0);
+      const int ngroups = get_num_groups(0);
+      const int wgid = get_group_id(0);
+
+      __global double* u = P + P_rows * (k + j) + k + j + 1;
+      __global double* v = V + V_rows * j + j;
+      double acc = 0;
+
+      for (int i = lid; i < P_rows - k - j - 1; i += LOCAL_SIZE_) {
+        double vi = v[i] + v[i + V_rows];
+        v[i] = vi;
+        acc += u[i] * vi;
+      }
+      __local double res_loc[LOCAL_SIZE_];
+      res_loc[lid] = acc;
+      barrier(CLK_LOCAL_MEM_FENCE);
+      for (int step = lsize / REDUCTION_STEP_SIZE; step > 0;
+           step /= REDUCTION_STEP_SIZE) {
+        if (lid < step) {
+          for (int i = 1; i < REDUCTION_STEP_SIZE; i++) {
+            res_loc[lid] += res_loc[lid + step * i];
+          }
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+      }
+      acc = res_loc[0] * 0.5;
+      for (int i = lid; i < P_rows - k - j - 1; i += LOCAL_SIZE_) {
+        v[i] -= acc * u[i];
+      }
+      if (gid == 0) {
+        P[P_rows * (k + j + 1) + k + j] -= *q / sqrt(2.) * u[0];
+      }
+    }
+    // \cond
+);
+// \endcond
+
+const kernel_cl<in_out_buffer, in_out_buffer, out_buffer, int, int, int, int>
+    tridiagonalization_householder("tridiagonalization_householder",
+                                   {tridiagonalization_householder_kernel_code},
+                                   {{"REDUCTION_STEP_SIZE", 4},
+                                    {"LOCAL_SIZE_", 1024}});
+
+const kernel_cl<in_buffer, in_buffer, out_buffer, out_buffer, int, int, int>
+    tridiagonalization_v1("tridiagonalization_v1",
```

Comment:
These are steps rather than versions right? So I would rename the kernels as step1, step2, etc.

Reply:
Yeah they are steps in calculation of vector `v`. I will rename them to remove the confusion.